### PR TITLE
Fix Hash(c10::Scalar), account for garbage data in union

### PR DESCRIPTION
**Pull request by wconstab**
Fixes #{issue number}

### CODE REVIEW

File: torch/csrc/lazy/core/hash.h [hunk 1]
```diff
@@ -74,7 +74,18 @@ static inline hash_t Hash(const c10::ScalarType& value) {
...
+  case c10::ScalarType::ComplexDouble:
+    return Hash(value.toComplexDouble());
+  case c10::ScalarType::Double:
+    return Hash(value.toDouble());
+  case c10::ScalarType::Long:
+    return Hash(value.toLong());
+  case c10::ScalarType::Bool:
+    return Hash(value.toBool());
+  default:
+    throw std::runtime_error("Unknown scalar type.");
```
Comment (alanwaketan) [inline torch/csrc/lazy/core/hash.h:87]: Maybe better to use TORCH_CHECK if we want users to know the error or TORCH_INTERNAL_ASSERT.

**Review by alanwaketan — APPROVED** [review]
LGTM.

File: torch/csrc/lazy/core/hash.h (hunk 1 above)
Comment (wconstab) [reply torch/csrc/lazy/core/hash.h:87]: i don't mind changing, but what's the difference actually?  

(I copied this exact throw from c10/Scalar.h, where it throws in `Scalar::type()`, so it's probably not too bad..)

Comment (alanwaketan) [reply torch/csrc/lazy/core/hash.h:87]: Those are just better (pretty print) wrappers for throwing exceptions, I think.

**Pull request merged**